stage0/src/sev.rs

```diff
@@ -26,16 +27,108 @@ use oak_sev_guest::{
     },
 };
 use spinning_top::Spinlock;
+use strum::FromRepr;
 use x86_64::{
     instructions::tlb,
-    registers::control::Cr3,
+    registers::{control::Cr3, model_specific::Msr},
     structures::paging::{
         frame::PhysFrameRange, page::AddressNotAligned, Page, PageSize, PageTable, PageTableFlags,
         PhysFrame, Size1GiB, Size2MiB, Size4KiB,
     },
     PhysAddr, VirtAddr,
 };
 
+/// The cache memory type used with MTRR.  We only use Write-Protect mode which the Linux kernel
+/// expects to be enabled by the firmware in order to enable SEV.
+#[repr(u8)]
+#[allow(dead_code)] // Remove if this is ever ported to a public crate.
+#[derive(FromRepr)]
+pub enum MemoryType {
+    UC = 0, // Uncacheable.
+    WC = 1, // Write-Combining.
+    WT = 4, // Writethrough	Reads.
+    WP = 5, // Write-Protect.
+    WB = 6, // Writeback.
+}
+
+impl TryFrom<u8> for MemoryType {
+    type Error = &'static str;
+    fn try_from(value: u8) -> Result<MemoryType, &'static str> {
+        match MemoryType::from_repr(value) {
+            Some(memory_type) => Ok(memory_type),
+            None => Err("invalid value for MemoryType"),
+        }
+    }
+}
+
+/// IA32_MTRR_DefType base model specific register.
+/// See https://wiki.osdev.org/MTRR for documentation.
+#[derive(Debug)]
+pub struct MTRRDefType;
+
+bitflags! {
+    /// Flags of the MTRRDefType Register.
+    pub struct MTRRDefTypeFlags: u64 {
+        /// Set this to enable MTRR.
+        const MTRR_ENABLE = 1 << 11;
+        /// Set to enable fixed-range support.
+        const FIXED_RANGE_ENABLE = 1 << 10;
+    }
+}
+
+impl MTRRDefType {
+    /// The underlying model specific register.
+    const MSR: Msr = Msr::new(0x2ff);
+
+    #[allow(dead_code)] // Remove if this is ever ported to a public crate.
+    pub fn read() -> (MTRRDefTypeFlags, MemoryType) {
+        let msr_value = Self::read_raw();
+        let memory_type:Result<MemoryType, &'static str>  = (msr_value as u8).try_into();
+        (
+            MTRRDefTypeFlags::from_bits_truncate(msr_value),
+            memory_type.unwrap(),
```

Comment:
In general prefer `expect`  instead of unwrap,  with a message describing what expectation failed.